File: python/common/trie.py
class Trie:
  def __init__(self):
    self.is_final = False
    self.children = [None] * 26

  def insert(self, word, i):
    if i == len(word):
      self.is_final = True
      return

    k = ord(word[i]) - ord('a')
    if self.children[k] == None:
      self.children[k] = Trie()

    self.children[k].insert(word, i + 1)

  def search(self, word, i):
    if i == len(word):
      return self.is_final

    k = ord(word[i]) - ord('a')

    if self.children[k] == None:
      return False
    return self.children[k].search(word, i + 1)

File: ruby/common/trie.rb
# Trie
class Trie
  attr_accessor :is_final, :children

  def initialize
    @is_final = false
    @children = Array.new(26)
  end

  def insert(word, i)
    if i == len(word)
      self.is_final = true
      return
    end

    k = word[i].ord - 'a'.ord
    children[k] = Trie.new if children[k].nil?

    children[k].insert(word, i + 1)
  end

  def search(word, i)
    return @is_final if i == word.length

    k = word[i].ord - 'a'.ord
    return false if @children[k].nil?

    @children[k].search(word, i + 1)
  end
end
